modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/QueryProcessor.java

```diff
@@ -53,4 +54,33 @@ CompletableFuture<AsyncSqlCursor<List<Object>>> querySingleAsync(
             String qry,
             Object... params
     );
+
+    /**
+     * Execute the multi-statement query with given schema name and parameters.
+     *
+     * @param properties User query properties. See {@link QueryProperty} for available properties.
+     * @param transactions Transactions facade.
+     * @param transaction A transaction to use for query execution. If null, an implicit transaction
+     *      will be started by provided transactions facade.
+     * @param qry Single statement SQL query.
+     * @param params Query parameters.
+     * @return Sql cursor.
+     *
+     * @throws IgniteException in case of an error.
+     */
+    CompletableFuture<AsyncSqlCursorIterator<List<Object>>> queryScriptAsync(
```

Comment:
You mean
```
**CompletableFuture**<AsyncIterator<AsyncSqlCursor<List<Object>>>> queryScriptAsync(...);
```
?
is it ok? :thinking:

Reply:
Should it prefetch the first element (current) similar to AsyncResultSet?

Reply:
hm, I see confusion... My suggestion was to bring `java.util.Iterator` to async world. With `java.util.Iterator` the first element is not available until you call `next`.

So, in order to preserve semantic of `java.util.Iterator`, we need only two operation: `hasNext` and `next`. Perhaps, we should proceed without `current` to avoid confusion

Reply:
Seems current() with prefetch adds complications for the user.
We do not have a method for checking that the iterator is empty (hasRowSet), so the user must, after receiving the iterator, not only call hasNext/fetchNext but also check that "current" exists or not.

Reply:
Sorry, I missed your last comment, don't mind my last one. Thank you.

Reply:
Thanks, added  `AsyncIterator` interface with  basic implementation `AsyncIteratorImpl`.